### PR TITLE
Draft: boltcards extension

### PR DESCRIPTION
**Pull request by iWarpBTC**
A draft (or rather a playground so far) for the bolt cards.

## Todo:
* [ ] Make sure to rebase to `main`
* [x] Apply `make format`
* [ ] Write tests for the extension
* [x] Add migrations of new extensions to `tools/conv.py`
* [x] Write `README.md` for extension

### CODE REVIEW

**Comment by Sparrowhawk1337** [comment]
This code could target any NXP card/tag/sticker/wristband/etc with SUN authentication, not just bolt cards.

Currently, these ones:

- NXP NTAG 223 DNA
- NXP NTAG 224 DNA
- NXP NTAG 424 DNA
- NXP NTAG 426Q

See here for an article on how to set up your own card.

[A self-sovereign contactless payment card](https://www.whitewolftech.com/articles/payment-card/)

**Comment by callebtc** [comment]
Thank you for the PR!

Please rebase to main and see if the checks succeed! It would be also great if you could write some tests for this extension. I've added a short list of todo's to your OP, I hope that's fine!

**Comment by callebtc** [comment]
How is this related to PR #725? Can it be considered a duplicate?

**Comment by iWarpBTC** [comment]
I would say no duplicity here. With PR #725 you can write a static (and most likely reusable) LNURLw on your tag. But some andvance nfc tags like NXP NTAG used by Bolt cards can give a unique link with every tap. If this link is pointed (idealy in acordance with [LUD-17](https://github.com/fiatjaf/lnurl-rfc/blob/luds/17.md)) to my extension (and some keys is set) my ext validate this link and response with a one time LNURLw. This is also called [Secure Unique NFC](https://mishka-scan.com/blog/secure-unique-nfc) and it's how CoinCorner use it.

**Comment by leesalminen** [comment]
@iWarpBTC is correct in their explanation. I actually have some Bolt Cards and generic NTAG424 devices en route to me to try out this extension :). Both solutions (simple LNURLw NFC tags and SUN authenticated NFC gags) are valuable additions.

**Comment by callebtc** [comment]
Thank you, got it! Really looking forward to this, sounds amazing. Wish I had one of those at hand!

**Comment by callebtc** [comment]
Sorry for the conflict. We recently moved `conv.py` to `tools/conv.py`. Needs one more rebase 😅

**Comment by callebtc** [comment]
Hi @iWarpBTC, all tests are green! Wanna go forward with this?

File: lnbits/extensions/boltcards/crud.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            withdraw,
+            file_key,
+            meta_key
+        )
+        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
+        """,
+        (
+            card_id,
+            wallet_id,
+            data.name,
```
Comment (leesalminen) [inline lnbits/extensions/boltcards/crud.py:29]: In my testing of this branch locally, I had to change `data.name` to `data.card_name`.

**Comment by leesalminen** [comment]
I've been testing this branch locally after getting a bunch of NXP NTAG 424 DNA in the mail in wristband & card forms. It works, w00!

Some testing notes that might require a code change:

- had to make a small code change in `crud.py`, listed in my comment above on this thread.
- if you edit a Card, close the modal *without* saving, and then click "Add Card", the form is pre-populated with the last edit.

Also, I started using the article provided by the OP, but I didn't get the desktop card reader in time and needed a way to program the NFC tags from an Android device. I was able to make the [TagWriter app by NXP work ](https://play.google.com/store/apps/details?id=com.nxp.nfc.tagwriter) work. Here's some steps on how to use:

- New Data Set > Link
- Set URI type to Custom URL
- URL should look like `lnurlw://YOUR_LNBITS_DOMAIN/boltcards/api/v1/scane?e=00000000000000000000000000000000&c=0000000000000000`
- click `Configure mirroring options`
- Select Card Type NTAG 424 DNA
- Check Enable SDM Mirroring
- Select SDM Meta Read Access Right to 01
- Check Enable UID Mirroring
- Check Enable Counter Mirroring
- Set SDM Counter Retrieval Key to 0E
- Set PICC Data Offset to immediately after `e=`
- Set Derivation Key for CMAC Calculation to 00
- Set SDM MAC Input Offset to immediately after `c=`
- Set SDM MAC Offset to immediately after `c=`
- Save & Write
- Scan with compatible Wallet

**Comment by iWarpBTC** [comment]
@leesalminen Thanks for the testing and thanks for the notes and the android setup guide! Let me know if there is something else that could be fixed or improved.

@callebtc Hello and sorry I didn't respond earlier, kind of overwhelmed this summer. Not sure if it's ready to get in. I still consider this rather a playground. For example from UX point of view the necessity of turning on an another extension (for lnurlw) and setting up a link there isn't cool. Wouldn't be better to include this lnurlw functionality inside this extension? But if you like it as it is we can go forward.

File: lnbits/extensions/boltcards/crud.py (hunk 1 above)
Comment (iWarpBTC) [reply lnbits/extensions/boltcards/crud.py]: Fixed. Thanks.

**Comment by iWarpBTC** [comment]
fucking hell this git

**Comment by leesalminen** [comment]
Sorry that happened to you @iWarpBTC :(. Luckily I had a local copy of your branch before you force-pushed a rebase (?). I re-created the branch by cherry picking your commits.

 I can take over the PR if you'd like (https://github.com/lnbits/lnbits-legend/pull/876). Or, if you want to continue just let me know and I can help you get the branch.

**Comment by iWarpBTC** [comment]
Totally cool if you take over the PR cause I suck with git. I didn't lose my work, only screw up the branch and the PR. I've actually done some work over the weekend. How can I merge it into the new PR now?

**Comment by leesalminen** [comment]
I usually use the GitHub CLI tool to do that. You can use it to check out my PR, then push up to my branch. I just gave you permission on the repo.

**Comment by leesalminen** [comment]
Here's a screenshot:

Screen Shot 2022-08-14 at 2 30 01 PM

**Pull request closed without merging**